flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSplitters.java

```diff
@@ -0,0 +1,110 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collection;
+
+import static org.apache.flink.connector.mongodb.common.utils.MongoConstants.ERROR_MESSAGE_FIELD;
+import static org.apache.flink.connector.mongodb.common.utils.MongoUtils.isCommandSucceed;
+
+/** To split collections of MongoDB to {@link MongoSourceSplit}s. */
+@Internal
+public class MongoSplitters implements Serializable, Closeable {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoSplitters.class);
+
+    private final MongoReadOptions readOptions;
+    private final boolean limitPushedDown;
+    private final MongoClient mongoClient;
+
+    public MongoSplitters(
+            MongoConnectionOptions connectionOptions,
+            MongoReadOptions readOptions,
+            boolean limitPushedDown) {
+        this.readOptions = readOptions;
+        this.limitPushedDown = limitPushedDown;
+        this.mongoClient = MongoClients.create(connectionOptions.getUri());
+    }
+
+    public Collection<MongoScanSourceSplit> split(MongoNamespace namespace) {
+        BsonDocument collStats = MongoUtils.collStats(mongoClient, namespace);
+        if (!isCommandSucceed(collStats)) {
+            LOG.error(
+                    "Execute command collStats failed: {}",
+                    collStats.getString(ERROR_MESSAGE_FIELD));
+            throw new IllegalStateException(String.format("Collection not found %s", namespace));
+        }
+
+        MongoSplitContext splitContext =
+                MongoSplitContext.of(readOptions, mongoClient, namespace, collStats);
+
+        if (limitPushedDown) {
+            LOG.info("Limit {} is applied, using single splitter", limitPushedDown);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
+
+        PartitionStrategy strategy = readOptions.getPartitionStrategy();
+        switch (strategy) {
+            case SINGLE:
+                return MongoSingleSplitter.INSTANCE.split(splitContext);
+            case SAMPLE:
+                return MongoSampleSplitter.INSTANCE.split(splitContext);
+            case SPLIT_VECTOR:
+                return MongoSplitVectorSplitter.INSTANCE.split(splitContext);
+            case SHARDED:
+                return MongoShardedSplitter.INSTANCE.split(splitContext);
+            case DEFAULT:
+            default:
+                return splitContext.isSharded()
+                        ? MongoShardedSplitter.INSTANCE.split(splitContext)
+                        : MongoSplitVectorSplitter.INSTANCE.split(splitContext);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (mongoClient != null) {
+            mongoClient.close();
+        }
+    }
+
+    /** Mongo Splitter to split a collection into multiple splits. */
+    public interface MongoSplitter {
```

Comment:
With all implementations being singletons this interface doesn't need to exist. They could all just be static methods.